File: scripts/fetch_wiki.py
import urllib.request
import re
import json

def fetch_wiki_winners(url, league):
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    html = urllib.request.urlopen(req).read().decode('utf-8')
    
    # We will just extract the table that contains the "Winning years"
    table_pattern = re.compile(r'<table class="wikitable[^>]*>.*?</table>', re.DOTALL)
    tables = table_pattern.findall(html)
    
    for table in tables:
        if 'Winning years' in table or 'Winning Seasons' in table or 'Years won' in table or 'Winning years' in table:
            rows = re.findall(r'<tr.*?>(.*?)</tr>', table, re.DOTALL)
            results = []
            for row in rows:
                cols = re.findall(r'<t[dh].*?>(.*?)</t[dh]>', row, re.DOTALL)
                if len(cols) >= 3:
                    # Strip html tags
                    clean_cols = [re.sub(r'<[^>]+>', '', c).strip() for c in cols]
                    club = clean_cols[0].replace('&#160;', '').replace('&amp;', '&').strip()
                    club = re.sub(r'\[.*?\]', '', club)
                    
                    # Usually winning years is in the 3rd or 4th column
                    years_col = clean_cols[2] if len(clean_cols[2]) > 5 else clean_cols[-1]
                    
                    years = re.findall(r'\b(18\d{2}(?:[/-]\d{2,4})?|19\d{2}(?:[/-]\d{2,4})?|20\d{2}(?:[/-]\d{2,4})?)\b', years_col)
                    if len(years) > 0 and club and club != 'Club' and club != 'Team':
                        results.append({'team': club, 'wins': years})
            if results:
                return results
    return []

print(json.dumps({
    'LL': fetch_wiki_winners("https://en.wikipedia.org/wiki/List_of_Spanish_football_champions", 'LL'),
    'BL': fetch_wiki_winners("https://en.wikipedia.org/wiki/List_of_German_football_champions", 'BL'),
    'L1': fetch_wiki_winners("https://en.wikipedia.org/wiki/List_of_French_football_champions", 'L1')
}, indent=2))

File: scripts/fetch_wiki.js
const fs = require('fs');

async function fetchWikiWinners(url, tableKeyword) {
  try {
    const res = await fetch(url, { headers: { 'User-Agent': 'Mozilla/5.0' } });
    const html = await res.text();

    const tablePattern = /<table class="wikitable[^>]*>([\s\S]*?)<\/table>/g;
    let match;
    const tables = [];
    while ((match = tablePattern.exec(html)) !== null) {
      tables.push(match[1]);
    }

    for (const table of tables) {
      if (table.includes(tableKeyword) || table.includes('Winning years')) {
        const rowPattern = /<tr.*?>([\s\S]*?)<\/tr>/g;
        let rowMatch;
        const results = [];
        while ((rowMatch = rowPattern.exec(table)) !== null) {
          const row = rowMatch[1];
          const colPattern = /<t[dh].*?>([\s\S]*?)<\/t[dh]>/g;
          let colMatch;
          const cols = [];
          while ((colMatch = colPattern.exec(row)) !== null) {
            cols.push(colMatch[1]);
          }

          if (cols.length >= 3) {
            const cleanCols = cols.map(c => c.replace(/<[^>]+>/g, '').replace(/&#160;/g, '').replace(/&amp;/g, '&').trim());
            let club = cleanCols[0].replace(/\[.*?\]/g, '').trim();
            if (club === '' || club === 'Club' || club === 'Team') continue;

            const yearsCol = cleanCols[2].length > 5 ? cleanCols[2] : cleanCols[cleanCols.length - 1];
            
            const yearPattern = /\b(18\d{2}(?:[/-]\d{2,4})?|19\d{2}(?:[/-]\d{2,4})?|20\d{2}(?:[/-]\d{2,4})?)\b/g;
            let yMatch;
            const years = [];
            while ((yMatch = yearPattern.exec(yearsCol)) !== null) {
              years.push(yMatch[1]);
            }

            if (years.length > 0) {
              results.push({ team: club, wins: years });
            }
          }
        }
        if (results.length > 0) return results;
      }
    }
    return [];
  } catch (e) {
    console.error(e);
    return [];
  }
}

async function run() {
  const data = {
    LL: await fetchWikiWinners("https://en.wikipedia.org/wiki/List_of_Spanish_football_champions", "Winning years"),
    BL: await fetchWikiWinners("https://en.wikipedia.org/wiki/List_of_German_football_champions", "Winning years"),
    L1: await fetchWikiWinners("https://en.wikipedia.org/wiki/List_of_French_football_champions", "Winning years"),
    PL: await fetchWikiWinners("https://en.wikipedia.org/wiki/List_of_English_football_champions", "Winning seasons")
  };
  console.log(JSON.stringify(data, null, 2));
}

run();
